library/src/main/kotlin/com/freeletics/coredux/ReduxStore.kt

```diff
@@ -2,116 +2,121 @@ package com.freeletics.coredux
- * the current State and an Action to compute a new State.
- * The new state will be emitted downstream to any listener interested in it.
- *
- * A ReduxStore observable never reaches onComplete(). If a error occurs in the [Reducer] or in any
- * side effect ([Throwable] has been thrown) then the ReduxStore reaches onError() as well and
- * according to the reactive stream specs the store cannot recover the error state.
+ * A [reduxStore] is a Kotlin coroutine based implementation of Redux and redux.js.org.
  *
  * @param initialState The initial state. This one will be emitted directly in onSubscribe()
- * @param sideEffects The sideEffects. See [SideEffect]
+ * @param stateReceiver function that receives new state updates
+ * @param sideEffects The sideEffects. See [SideEffect].
  * @param reducer The reducer.  See [Reducer].
  * @param S The type of the State
  * @param A The type of the Actions
+ *
+ * @return [ActionDispatcher] that should be use to dispatch new actions to [reduxStore]
  */
-@UseExperimental(ExperimentalCoroutinesApi::class, ObsoleteCoroutinesApi::class)
-fun <S: Any, A: Any> ReceiveChannel<A>.reduxStore(
+@UseExperimental(ExperimentalCoroutinesApi::class)
+fun <S: Any, A: Any> CoroutineScope.reduxStore(
     initialState: S,
-    sideEffects: List<SideEffect<S, A>>,
-    coroutineScope: CoroutineScope = GlobalScope,
+    stateReceiver: StateReceiver<S>,
+    sideEffects: List<SideEffect<S, A>> = emptyList(),
     reducer: Reducer<S, A>
-): ReceiveChannel<S> {
-    val output = Channel<S>()
-    val actionsChannel = BroadcastChannel<A>(1 + sideEffects.size)
-    var currentState = initialState
-
-    output.invokeOnClose {
-        this@reduxStore.cancel()
-        actionsChannel.close()
+): ActionDispatcher<A> {
+    val actionsReducerChannel = Channel<A>(Channel.UNLIMITED)
+    val actionsSideEffectsChannel = BroadcastChannel<A>(CONFLATED)
+
+    val actionDispatcher: ActionDispatcher<A> = { action ->
+        if (actionsReducerChannel.isClosedForSend) throw IllegalStateException("CoroutineScope is cancelled")
+
+        if (!actionsReducerChannel.offer(action)) {
+            throw IllegalStateException("Input actions overflow - buffer is full")
+        }
     }
 
-    // Sending initial state
-    coroutineScope.launch {
-        output.send(currentState)
+    coroutineContext[Job]?.invokeOnCompletion {
+        actionsReducerChannel.close(it)
+        actionsSideEffectsChannel.close(it)
     }
 
     // Starting reducer coroutine
-    coroutineScope.launch {
+    launch {
+        var currentState = initialState
+
+        // Sending initial state
+        stateReceiver(currentState)
+
+        // Starting side-effects coroutines
+        sideEffects.forEach { sideEffect ->
+            with (sideEffect) {
+                start(
+                    actionsSideEffectsChannel.openSubscription(),
+                    { currentState },
```

Comment:
As I understand it, this is a reference to `currentState`. The `sideEffect` is running in a Coroutine. Which version of the state does the `sideEffect` receive?